### PR TITLE
std: Add impl of FnOnce to AssertRecoverSafe

### PR DESCRIPTION
**Pull request by alexcrichton**
This was originally intended, but forgot to land by accident!

cc #27719

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @aturon 

cc @sfackler, @wycats, @nikomatsakis

File: src/libstd/panic.rs [hunk 1]
```diff
@@ -144,10 +147,28 @@ pub trait RefRecoverSafe {}
 /// // });
 ///
 /// // This, however, will compile due to the `AssertRecoverSafe` wrapper
+/// let result = panic::recover(AssertRecoverSafe::new(|| {
+///     variable += 3;
+/// }));
+/// // ...
+/// ```
+///
+/// Sometimes this behavior may be a bit too coarse grained, however, and
```
Comment (nikomatsakis) [inline src/libstd/panic.rs]: I think this comment is kind of vague. I would maybe say something like:

Wrapping the entire closure amounts to a blanket assertion that all variables are recover safe. This has the downside that if new captures are added in the future, they will also be considered recover safe. Therefore, you may prefer to just wrap individual capture, as shown below. This is more annotation, but it ensures that if a new upvar is added which is not recover safe, you will get a compilation error at that time, which will allow you to consider whether that new capture in fact represent a bug or not.

Comment (alexcrichton) [inline src/libstd/panic.rs]: Thanks! Updated with this text.

**Comment by glaebhoerl** [comment]
Is there a reason not to have `Fn` and `FnMut` too?

**Comment by sfackler** [comment]
`FnOnce` is a supertrait of `Fn` and `FnMut`, so those come along for free.

**Comment by Stebalien** [comment]
@sfackler 

> FnOnce is a supertrait of Fn and FnMut,

yes

> so those come along for free.

no.

`Fn` implies `FnMut` implies `FnOnce`, not the other way arround. If you only implement `FnOnce`, you can only call call `AssertRecoverSafe` once.

**Comment by aturon** [comment]
We could certainly add `Fn` and `FnMut` impls, but there's not much reason to: `Recover` takes a `FnOnce`.

**Comment by glaebhoerl** [comment]
Yeah I hadn't absorbed that the *only* place you'd want this is on the argument to `recover`. (I suppose other HOFs bounding their arguments by `RecoverSafe` isn't likely.)

**Pull request merged**